Async ProtocolStreamWriter

MadHed opened the pull request:
Here is my proposal for an async ProtocolStreamWriter.

Currently 1 test fails, because it expects to write all data immediately, which is now not the case anymore.

Maybe this will fix https://github.com/felixfbecker/php-language-server/pull/110#issuecomment-255836119

felixfbecker commented:
I think `write()` should return a promise that resolves when the data has been written.

What about this case:
 - Loop says "hey the stream is ready, you can write stuff" - buffer is empty though
 - `write()` is called
 - when will the callback be called?

MadHed commented:
> I think write() should return a promise that resolves when the data has been written.

Good Idea.

> when will the callback be called by the Loop?

The callback will be called every time `stream_select()` is called... which I just found out. When doing nothing the server just sits there busy-waiting until a message arrives.

felixfbecker commented:
> The callback will be called every time stream_select() is called... which I just found out. When doing nothing the server just sits there busy-waiting until a message arrives, consuming cpu cycles.

"Until a message arrives" is for the read stream, what about the write case here? Can we guarantee that a message will eventually be written out?

MadHed commented:
On my machine (Windows) the write callback is called all the time (`stream_select` never blocks). Which is actually a bit worrysome because without doing anything the php process soncumes about 14% cpu time.

MadHed commented:
I think the correct procedure would be to call `Loop\addWriteStream()` when `write()` is called and `Loop\removeWriteStream()` when the last message was compeltely sent.

MadHed commented:
The last commit must be handled with care! Somehow vscode completely hangs after parsing about 900 files and I have no idea why. It seems to be some kind of timing problem, because after adding logging calls the bug disappears.

kaloyan-raev commented on src/ProtocolStreamWriter.php:
```diff
@@ -26,22 +38,57 @@ public function __construct($output)
+        error_clear_last();
+        $bytesWritten = @fwrite($this->output, $message);
+
+        if ($bytesWritten === false) {
```
This clause would never be satisfied. If using `@`, `fwrite` will return 0 on error instead of FALSE. The whole error handling needs to be removed.

kaloyan-raev replied:
I was not completely correct, but still this clause would be never satisfied.

`fwrite` returns FALSE only if invalid arguments are passed. `fwrite` will return 0 (or even higher value) on all other runtime errors, like "broken pipe" or "non-blocking operation not possible".

Please read this comment: http://php.net/manual/en/function.fwrite.php#96951

MadHed replied:
Yeah, already forgot about this. Should we just ignore it then?

felixfbecker replied:
Yep, as it is not a runtime error.

kaloyan-raev commented on src/ProtocolStreamWriter.php:
```diff
@@ -26,22 +38,57 @@ public function __construct($output)
+
+            $promise->fulfill();
+        } else {
+            $this->messages[0]['message'] = $message;
```
What does happen with the promise in this case?

felixfbecker replied:
> BTW, is there any part of the code that relies on these promises? I don't find any.

No, not yet because we don't have any functions that have async return values. The only async function is `indexProject()`, but it doesn't have a return value. I was planning to use more promises though, in particular here in the ProtocolStreamWriter and anything that involves file IO

kaloyan-raev commented:
I've done some testing of this PR. It works OK for me. I did not have any hangs.

If I add [this check](https://github.com/felixfbecker/php-language-server/blob/83afa0c1b8f779d339d2034e4e6fdc154762fdb3/src/ProtocolStreamReader.php#L33-L35), I also have the issue with orphan php processes resolved.

MadHed commented:
Removed the `false` check as per https://github.com/felixfbecker/php-language-server/pull/112#pullrequestreview-5589368

Also modified the writer to always try to write all pending messages in one tick.
The indexer now runs through without problems on Windows.

IMO another good addition would be to add a `writeUrgent()` which puts the message in front of the pending messages list. But this can be added later if needed. This would allow us to send unimportant messages like Diagnostics and LogMessage with a low priority while serving Symbol requests and the like immediately.

felixfbecker commented:
Why should it try to write all pending messages in one tick?

MadHed commented:
> Why should it try to write all pending messages in one tick?

The freeze I encountered had to do something with the amount of queued messages. Still not sure what exactly happened there. But since we are sending at least 2 messages per parsed, file the queue would grow constantly and eventually vscode window would freeze up. This fixes it.

kaloyan-raev commented:
My feeling is that this would make the language server more responsive during indexing. I like the idea for `writeUrgent()`.

MadHed commented:
I think writing all pending messages is correct for responsiveness reasons. If we were to write only a single message per tick the queue would quickly fill up while we are processing other server requests. Since we are not constrained by bandwidth this would artificially increase our response time. The writer still remains nonblocking because it yields as soon as a message could not be sent in full.

kaloyan-raev commented on src/ProtocolStreamWriter.php:
```diff
@@ -59,36 +59,33 @@ public function write(Message $msg)
 
-        error_clear_last();
-        $bytesWritten = @fwrite($this->output, $message);
+            error_clear_last();
```
Is this call still necessary?

felixfbecker commented:
Any reason why this much simpler version would not work;
```php
$data = (string)$msg;
        $msgSize = strlen($data);
        $writeData = function () use (&$data, $msgSize, &$writeData) {
            $bytesWritten = @fwrite($this->output, $data);
            if ($bytesWritten < $msgSize) {
                $data = substr($data, $bytesWritten);
            } else {
                Loop\removeWriteStream($this->output, $writeData);
            }
        };
        Loop\addWriteStream($this->output, $writeData);
```
?

MadHed commented:
@felixfbecker Do you mean for `write()`?
The Sabre event loop only allows a single callback per stream. Also the messages still have to be sent one after another, otherwise they would become mixed up.

felixfbecker commented:
I think I am fine with this. The only question left open is about the call sites - before a call would block, so now the methods that call `write()` need to return the promise too.

felixfbecker commented:
The only places though where this is used currently is the client, and there is makes more sense to actually return a promise for the response. Not sure how to handle that yet.

MadHed commented:
> The only places though where this is used currently is the client, and there is makes more sense to actually return a promise for the response. Not sure how to handle that yet.

Yeah, I think it would make more sense at a higher abstraction level where LSP requests and responses are handled asynchronously via promises.
